### ISSUE TITLE
Accessible Authentication (3.3.7) far too fundamentally restrictive for a Level A

### ISSUE THREAD

**Issue by patrickhlauke** [w3c/wcag#1364]
While it's true that the SC's understanding https://www.w3.org/WAI/WCAG22/Understanding/accessible-authentication.html lists many technical options that would pass (though the first example of "properly marked up" email and password fields is odd - see #1363), fundamentally this SC would, overnight, make wast swathes of sites that use authentication fail. In particular (depending on the discussion on #1359) if established security practices like 2FA are not allowed / need to provide an alternative.

As I currently understand it, this SC would force even the smallest of websites to essentially make changes to their entire authentication architecture to provide one of the suggested methods. Yes, they're all technically feasible of course, but if I just own a small website with a basic username/password login, and a "reset your password" type functionality, it would seem that I'd have to start delving into far more complex authentication solutions. This seems unreasonable for a Level A requirement.

Also noting that if indeed only the solutions suggested in the understanding are valid ways to pass, my current bank (which requires a 2FA code to be copied from their app/typed into the web-based login field), github access (which uses a 2FA code again that needs to be copy/pasted out of 1password), and even w3c's own member access (which does offer a password recovery option, but does not send a link to login directly...you still have to then use the username/password they resent you to log in) will fail.

I'd say that either more options need to be mentioned/explained as a suitable way to pass this SC (does a recovery that emails you your username/password work? at least THAT would cover many bases), or this should be moved to a different Level (AA or AAA) due to the high technical burden this imposes on the vast majority of sites on the current web.

As an aside, I know initially the ideas here were mentioned/reviewed by security groups, and apparently they had no issues with it. I'd suggest that this be reviewed again, making it clear what the implications of this SC would actually be (essentially disallowing 2FA unless there's some other alternative listed in the understanding doc, and that sites that don't implement these will fail / be "illegal").

**Comment by patrickhlauke** [comment]
also cross-referencing https://github.com/w3c/wcag/issues/1358 which has similar concerns (though much better expressed) about this actually leading to less secure options being used (if an author/dev is not able to implement the stronger things like WebAuthN etc, having to abandon things like 2FA or similar)

**Comment by AidanA11y** [comment]
> initially the ideas here were mentioned/**reviewed by security groups,** and apparently they had no issues with it. I'd **suggest that this be reviewed again, making it clear what the implications of this SC would actually be (essentially disallowing 2FA** unless there's some other alternative 

I agree.

**Comment by alastc** [comment]
Going back to the WCAG 2.1 timeframe, this was [my main objection](https://lists.w3.org/Archives/Public/w3c-wai-gl/2017AprJun/1172.html). This SC needs to scale to both the small scale (personal blog) and large (email provider). Pre-webauthn it wasn't possible, but that has changed.

At the small-end of the scale we can use:
-	Simple username (or email) and password, assuming you don’t block copy-paste (and meet 1.3.5 but that’s separate).
-	‘Magic email’ to login without username password.
-	Email-reset loops.

Getting slightly larger, if you are implementing a second factor then you really want to use something provided, rolling your own is positively dangerous. In which case you have multiple providers, including via WebauthN. That basically is the standardised route to out-of-browser mechanisms.
That basically off-loads the issue to the user’s provider, you can use whatever mechanism you have setup. E.g. I can use Windows helo, in which case I type in a 4-digit code. You could also use face-ID, or whatever you use with your computer/device.

Note that WebauthN takes over the whole authentication process, it is not just a 2nd factor.

You also have the option to use a 3rd party via Oauth, or a dedicated provider like Duo.com (which I now use due to having a particular type of W3C account).

If you are a large provider (like an international email provider) who is rolling their own, then you have various options, and most are moving away from the time-based codes. Those time-based codes are not being made illegal, but the requirement is to provide another way.
Microsoft, Google & dedicated providers like Duo.com provide non-transcribing options. The general setup is that you start the login process on the site, and it tells you to go to the app, and it asks if you want to allow the login. I use at least 3 that I can think of, my favourite is the MS iOS app which pops up the message on my watch and I can ok it there.

> initially the ideas here were mentioned/reviewed by security groups, and apparently they had no issues with it. I'd suggest that this be reviewed again, making it clear what the implications of this SC would actually be (essentially disallowing 2FA unless there's some other alternative

We will have to go through a security review, I think that process has started (although it isn't quick).

**Comment by alastc** [comment]
Draft response for discussion with the group:
<hr>
From the initial comment there appear to be some misunderstandings of what will pass/fail the criterion, PR #1499 has been added to make updates. The initial one is an example of resetting a password via email. The other examples already show how 2FA can be accomplished.

With the [comment above](#issuecomment-693379705) and the fact this SC will go through a security review at W3C, we think that addresses the general point about the restrictiveness of the SC.

**Comment by alastc** [comment]
Hi everyone,

The group agreed the draft response, with the caveat that it will go through a W3C security review and we are encouraging everyone to run it past their clients / security teams for any specific issues.

**Comment by mraccess77** [comment]
So it seems like copy/paste is acceptable - but only if the fields meet SC 1.3.5.  You can meet 1.3.5 with micro data rather than autocomplete but seems that we really want the field to be autopopulated.  But it's unclear if the auto populate we are talking about is auto populate from text message or about saving for next time after pasting in updated password/username.  1.3.5 isn't about auto populating but instead about field purpose.  So are we really asking to validate auto populate and is that automatically from the email/text message or only remembering for next time?

**Comment by patrickhlauke** [comment]
as an example then, my current bank requires the use of a either a physical security key (a tiny physical calculator type jobby) or a digital security key for iOS/Android. in both cases, the security key is separate from the computer, and requires manually copying a 6 digit code from the key and into a text field on the website. there is no alternative to this (no fancy new WebAuthN or similar, and no way to avoid/circumvent the second factor check). I assume my bank would fail under this SC, correct?

**Comment by alastc** [comment]
@mraccess77 I think you are discussing #1359?

@patrickhlauke correct & right, unless they provide another mechanism. 

It probably won't shock you to hear that Banks don't tend to be on the forefront of authentication practices, but take a very conservative approach. This is a good 'both sides' article on [bank authentication practices](https://www.troyhunt.com/banks-arbitrary-password-restrictions-and-why-they-dont-matter/).

My personal bank has a (failing) login on the desktop that requires you to select characters from a previously invented password. However, on the app it uses faceID (on iOS).

My work bank (a different mainstream bank) also allows for faceID on the app. Niether have done away with passwords, but for repeat authentication you don't need one.

FaceID hasn't been around that long, perhaps with a little encouragement they can apply the same business and security logic to desktop with WebAuthN (or something else) and allow more customers to bank online?

Although, I wouldn't expect that in the next couple of years. It will take a while for 2.2 to become known & widely used, particularly by the more conservative sectors.
